File: 1124-longest-well-performing-interval/1124-longest-well-performing-interval.py
class Solution:
    def longestWPI(self, hours: List[int]) -> int:
        seen = {}
        source = ans = 0
        for i, hour in enumerate(hours):
            source += 1 if hour > 8 else -1
            if source > 0:
                ans = i + 1
            if source not in seen:
                seen[source] = i
            if source - 1 in seen:
                ans = max(ans, i - seen[source - 1])
        return ans

File: 1124-longest-well-performing-interval/1124-longest-well-performing-interval.js
/**
 * @param {number[]} hours
 * @return {number}
 */
var longestWPI = function(hours) {
  const map = {};
  let [ans, source] = [0, 0];
  
  for (let i = 0; i < hours.length; i++) {
    source += hours[i] > 8 ? 1 : -1;
    if (source > 0) ans = i + 1;
    if (!(source in map)) map[source] = i;
    if (source - 1 in map) ans = Math.max(ans, i - map[source - 1]);
  }
  
  return ans;
};